### DISCUSSION TITLE
How to switch urlencode params for json datasource to use %20 for spaces instead of +?

### DISCUSSION THREAD

**Discussion by veshus** [getredash/redash#6610] [Q&A]
How to switch urlencode params for json datasource to use %20 for spaces instead of +?

**Comment by guidopetri** [comment]
Could you describe more of what you're trying to do? I'm not sure exactly what you mean. You should be able to use `%20` if you want to, or use the `params` key and use plain spaces instead of `+`.

**Comment by veshus** [comment]
I have an API that treats space as %20 while url decode.
When I create a JSON datasource with params like this:

```
?Начало периода=2023-01-01Т00:00:00&Конец периода=2023-12-31T23:59:59
```

and It must encode to
``` 
%3F%D0%9D%D0%B0%D1%87%D0%B0%D0%BB%D0%BE%20%D0%BF%D0%B5%D1%80%D0%B8%D0%BE%D0%B4%D0%B0%3D2023-01-01%D0%A200%3A00%3A00%26%D0%9A%D0%BE%D0%BD%D0%B5%D1%86%20%D0%BF%D0%B5%D1%80%D0%B8%D0%BE%D0%B4%D0%B0%3D2023-12-31T23%3A59%3A59
```
but not as 
![image](https://github.com/getredash/redash/assets/33095952/68a0e1ed-11f5-411a-812f-c0867d172574)
like in default behaviour

**Comment by guidopetri** [comment]
Can you post the query you're using in redash to get that request?

**Comment by veshus** [comment]
It's looks like this

![image](https://github.com/getredash/redash/assets/33095952/9de7fc5a-0211-416f-bf26-3bfbc33a92fc)

**Reply by guidopetri** [reply]
I see, thanks for explaining. Unfortunately, [as far as I can tell](https://stackoverflow.com/a/29948396), this is the correct behavior - any spaces in the params should be encoded with a `+`, and any spaces before the params in the path should be encoded with a `%20`. Since this is a param, I'd say this is expected behavior, and thus an issue with the API and not redash.

Can you try replacing the spaces manually with `%20` and seeing if that works for you?

**Comment by veshus** [comment]
Absolutely. But it's legacy application and we can't change it's behaviour to match RFC, but we need data from it :(
If I change <space> to %20 myself, after urlencode it will be %2520, so the only way is to change the default behaviour of redash.
Now I tried to change it globally by adding ENV switch plus quote changer to the redash/utils/requests_session.py

**Reply by guidopetri** [reply]
Hmm that makes sense. I think what I'd prefer personally is to treat the parameter names as raw strings instead - that way you could use `%20` without it converting to `%2520`. What do you think of that alternative?

If that still doesn't fit your needs, then I think we can add the env var, but if we can help it, I'd prefer not to add something that explicitly doesn't support the standard - that might prove very confusing to someone in the future.

**Comment by veshus** [comment]
I'm not sure about parameters as raw strings, I need some tests on the datasource system.
On the other hand, env var and the [proposed PR](https://github.com/getredash/redash/pull/6621) looks transparent and safe because it switches the whole system to "legacy" mode.

**Reply by guidopetri** [reply]
Right, but it explicitly doesn't follow the standard and could be confusing in the future. So I don't think we'd want to merge it in if we can find a different solution.

**Comment by veshus** [comment]
Is it any standard way to REST query without encoding parameters at all?
It looks like out datasource accept raw UTF-8 strings

**Reply by guidopetri** [reply]
No, I don't think that's the standard, but I think that would be a viable option in this case if we were to set raw strings e.g. via a `r` prefix to the string. I would prefer a PR like that than the existing PR.
